### PR TITLE
add golangci lint to travis

### PR DESCRIPTION
**Pull request by draveness**
Related to #777

### CODE REVIEW

File: pkg/features/features.go [hunk 1]
```diff
@@ -54,7 +54,10 @@ const (
 )
 
 func init() {
-	utilfeature.DefaultFeatureGate.Add(defaultFederationFeatureGates)
+	if err := utilfeature.DefaultFeatureGate.Add(defaultFederationFeatureGates); err != nil {
+		// TODO:
```
Comment (draveness) [inline pkg/features/features.go]: What should I do when `Add` return error

File: pkg/kubefed2/disable.go [hunk 2]
```diff
@@ -65,6 +65,7 @@ type disableType struct {
 	disableTypeOptions
 }
 
+// nolint[:structcheck]
```
Comment (draveness) [inline pkg/kubefed2/disable.go]: is this a good approach to disable false positive, another option is to use a `golangci.yml` file to disable the structcheck in the file and `enable.go[1]`

File: pkg/kubefed2/enable/enable.go [hunk 3]
```diff
@@ -72,6 +72,7 @@ type enableType struct {
 	enableTypeOptions
 }
 
+// nolint[:structcheck]
```
Comment (draveness) [inline pkg/kubefed2/enable/enable.go]: [1]

File: pkg/kubefed2/disable.go (hunk 2 above)
Comment (marun) [reply pkg/kubefed2/disable.go]: What is golint complaining about in this instance?

File: pkg/kubefed2/enable/enable.go [hunk 4]
```diff
@@ -227,10 +228,14 @@ func GetResources(config *rest.Config, enableTypeDirective *EnableTypeDirective)
 // already been enabled for federation.  This would likely involve
 // updating the version of the target type and the validation of the schema.
 func CreateResources(cmdOut io.Writer, config *rest.Config, resources *typeResources, namespace string) error {
-	write := func(data string) {
+	write := func(data string) error {
```
Comment (marun) [inline pkg/kubefed2/enable/enable.go]: This is intended to be a braindead helper method.  It should not return an error and necessitate that every client checks for an error in turn.  Panicking should be sufficient.

File: pkg/kubefed2/enable/enable.go [hunk 5]
```diff
@@ -334,17 +344,18 @@ func GenerateTypeConfigForTarget(apiResource metav1.APIResource, enableTypeDirec
...
 	return CrdForAPIResource(typeConfig.GetFederatedType(), schema, shortNames)
 }
 
 func writeObjectsToYAML(objects []pkgruntime.Object, w io.Writer) error {
 	for _, obj := range objects {
-		w.Write([]byte("---\n"))
-		err := writeObjectToYAML(obj, w)
-		if err != nil {
+		if _, err := w.Write([]byte("---\n")); err != nil {
+			return err
```
Comment (marun) [inline pkg/kubefed2/enable/enable.go]: The error should be wrapped with the same message as line 359.

File: test/e2e/crud.go [hunk 6]
```diff
@@ -92,7 +92,12 @@ var _ = Describe("Federated", func() {
 					tl.Fatalf("Failed to create unlabeled resource in cluster %q: %v", clusterName, err)
 				}
 				clusterClient := genericclient.NewForConfigOrDie(clusterConfig)
-				defer clusterClient.Delete(context.TODO(), unlabeledObj, unlabeledObj.GetNamespace(), unlabeledObj.GetName())
+				defer func() {
+					err := clusterClient.Delete(context.TODO(), unlabeledObj, unlabeledObj.GetNamespace(), unlabeledObj.GetName())
+					if err != nil {
+						tl.Fatalf("Unexpected error: %v", err)
```
Comment (marun) [inline test/e2e/crud.go]: Will `Fatalf` allow other deferred calls to execute?

File: test/e2e/crud.go [hunk 7]
```diff
@@ -111,11 +116,16 @@ var _ = Describe("Federated", func() {
...
+					err := hostClient.Delete(context.TODO(), createdObj, createdObj.GetNamespace(), createdObj.GetName())
+					if err != nil {
+						tl.Fatalf("Unexpected error: %v", err)
+					}
+				}()
 
 				waitDuration := 10 * time.Second // Arbitrary amount of time to wait for deletion
 				By(fmt.Sprintf("Checking that the unlabeled resource is not deleted within %v", waitDuration))
-				wait.PollImmediate(framework.PollInterval, waitDuration, func() (bool, error) {
+				err = wait.PollImmediate(framework.PollInterval, waitDuration, func() (bool, error) {
```
Comment (marun) [inline test/e2e/crud.go]: Ignoring this error was intentional, as timing out is success, and it should continue to be ignored (e.g. `_ = wait.PollImmediate`).

File: test/e2e/framework/unmanaged.go [hunk 8]
```diff
@@ -213,8 +212,8 @@ func (f *UnmanagedFramework) ClusterDynamicClients(apiResource *metav1.APIResour
 		// Check if this cluster is the same name as the host cluster name to
 		// make it the primary cluster.
 		testClusters[clusterName] = common.TestCluster{
-			clusterConfig,
-			client,
+			TestClusterConfig: clusterConfig,
```
Comment (marun) [inline test/e2e/framework/unmanaged.go]: What rule is this violating exactly?

File: pkg/controller/federatedcluster/clusterclient.go [hunk 9]
```diff
@@ -125,21 +125,24 @@ func (self *ClusterClient) GetClusterZones() (zone, region string, err error) {
...
-			region, err = getRegionNameForNode(node)
-			if err != nil {
-				return "", "", err
-			}
-		}
-		// TODO: Optimize this flow. All nodes will have the same zone label.
-		// So just considering first node for now.
-		break
+
+	if len(nodes.Items) <= 0 {
```
Comment (marun) [inline pkg/controller/federatedcluster/clusterclient.go]: This code is being rewritten in #797 - please revert.

File: pkg/features/features.go (hunk 1 above)
Comment (marun) [reply pkg/features/features.go]: This is a pretty low value fix, I suggest leaving it for now.

File: pkg/kubefed2/unjoin.go [hunk 10]
```diff
@@ -285,14 +285,12 @@ func deleteFederatedClusterAndSecret(hostClientset kubeclient.Interface, client
 func deleteRBACResources(unjoiningClusterClientset kubeclient.Interface,
 	namespace, unjoiningClusterName, hostClusterName string, dryRun bool) bool {
 
-	var deletionSucceeded = true
```
Comment (marun) [inline pkg/kubefed2/unjoin.go]: What is golint complaining about in this instance?

File: test/e2e/framework/framework.go [hunk 11]
```diff
@@ -306,7 +306,7 @@ func createNamespace(client kubeclientset.Interface, baseName string) (string, e
 	// TODO(marun) should all api calls be made 'robustly'?
 	var namespaceName string
 	if err := wait.PollImmediate(PollInterval, TestContext.SingleCallTimeout, func() (bool, error) {
-		namespace, err := client.Core().Namespaces().Create(namespaceObj)
+		namespace, err := client.CoreV1().Namespaces().Create(namespaceObj)
```
Comment (marun) [inline test/e2e/framework/framework.go]: I get that V1 should be preferred, but how does golint know that?

File: test/e2e/crud.go (hunk 7 above)
Comment (marun) [reply test/e2e/crud.go]: Note that this change is responsible for the CI failure.

File: pkg/kubefed2/disable.go (hunk 2 above)
Comment (draveness) [reply pkg/kubefed2/disable.go]: some variables in the struct were unused. Excluded with a golangci lint file

File: test/e2e/crud.go (hunk 6 above)
Comment (draveness) [reply test/e2e/crud.go]: Yes, I have tested it.

File: pkg/kubefed2/unjoin.go (hunk 10 above)
Comment (draveness) [reply pkg/kubefed2/unjoin.go]:
```
ineffectual assignment to `deletionSucceeded`
```

the statement to initialzie `deletionSucceeded` var is not useful, which would be overwritten after calling `deleteClusterRoleAndBinding`

File: test/e2e/framework/unmanaged.go (hunk 8 above)
Comment (draveness) [reply test/e2e/framework/unmanaged.go]: Use the style above to initialise structs would reduce the possibility of unexpected initialises error, suppose we have something like:

```go
type Test struct {
  a int
  b int
}
```

Initialise the struct above use `Test{1, 2}` may lead to problems when the order of vars changes in the `Test` struct.

And there are only two places which use this approach to initialise the struct. It's better to maintain consistency in the project.

File: test/e2e/framework/framework.go (hunk 11 above)
Comment (draveness) [reply test/e2e/framework/framework.go]: The function gets a deprecated comment.

```go
	// Deprecated: please explicitly pick a version if possible.
	Core() corev1.CoreV1Interface
```

File: pkg/kubefed2/unjoin.go (hunk 10 above)
Comment (marun) [reply pkg/kubefed2/unjoin.go]: I would prefer that this check be disabled.  An 'ineffectual assignment' may well be useful for communicating intent to future maintainers even if it has no value to the golang runtime.

File: test/e2e/framework/unmanaged.go (hunk 8 above)
Comment (marun) [reply test/e2e/framework/unmanaged.go]: I get the point of the check, but I don't think it's worth enforcing at this time.  Please disable it for now.

File: pkg/kubefed2/disable.go (hunk 2 above)
Comment (marun) [reply pkg/kubefed2/disable.go]: I'm confused.  If they were truly unused, why not just remove them?  Or if they are used, and golint isn't smart enough to figure that out, I'm not sure the check is worth enabling.

File: pkg/features/features.go (hunk 1 above)
Comment (marun) [reply pkg/features/features.go]: I realize I was confused about what you were asking.  I think `glog.Fatalf` is likely the right answer.

File: pkg/controller/util/planner/planner.go [hunk 12]
```diff
@@ -71,25 +72,41 @@ func (p *Planner) Plan(availableClusters []string, currentReplicaCount map[strin
...
-		}
+		}, nil
 	}
 
 	for _, cluster := range availableClusters {
 		if localRSP, found := p.preferences.Spec.Clusters[cluster]; found {
-			preferences = append(preferences, named(cluster, localRSP))
+			preference, err := named(cluster, localRSP)
+			if err != nil {
+				fmt.Printf("Unexpected err: %s\n", err.Error())
```
Comment (marun) [inline pkg/controller/util/planner/planner.go]: `glog.Errorf` should be used instead of `Printf`

File: pkg/controller/util/planner/planner.go [hunk 13]
```diff
@@ -71,25 +72,41 @@ func (p *Planner) Plan(availableClusters []string, currentReplicaCount map[strin
...
+				continue
+			}
+
+			preferences = append(preferences, preference)
 		} else {
 			if localRSP, found := p.preferences.Spec.Clusters["*"]; found {
-				preferences = append(preferences, named(cluster, localRSP))
+				preference, err := named(cluster, localRSP)
+				if err != nil {
+					fmt.Printf("Unexpected err: %s\n", err.Error())
```
Comment (marun) [inline pkg/controller/util/planner/planner.go]: ditto

File: pkg/kubefed2/disable.go [hunk 14]
```diff
@@ -159,9 +159,11 @@ func DisableFederation(cmdOut io.Writer, config *rest.Config, typeConfigName ctl
 	}
 
 	write := func(data string) {
-		if cmdOut != nil {
-			cmdOut.Write([]byte(data))
+		if cmdOut == nil {
+			return
 		}
+
+		_, _ = cmdOut.Write([]byte(data))
```
Comment (marun) [inline pkg/kubefed2/disable.go]: I apologize for not being clear.  The error should not be ignored.  It should be logged with `glog.Fatalf`.

File: pkg/kubefed2/enable/enable.go [hunk 15]
```diff
@@ -229,7 +229,9 @@ func GetResources(config *rest.Config, enableTypeDirective *EnableTypeDirective)
 func CreateResources(cmdOut io.Writer, config *rest.Config, resources *typeResources, namespace string) error {
 	write := func(data string) {
 		if cmdOut != nil {
-			cmdOut.Write([]byte(data))
+			if _, err := cmdOut.Write([]byte(data)); err != nil {
+				panic(err)
```
Comment (marun) [inline pkg/kubefed2/enable/enable.go]: Same comment as for disable -`glog.Fatalf`

File: pkg/kubefed2/enable/enable.go [hunk 16]
```diff
@@ -255,6 +257,7 @@ func CreateResources(cmdOut io.Writer, config *rest.Config, resources *typeResou
 		if err != nil {
 			return errors.Wrapf(err, "Error creating CRD %q", resources.CRD.Name)
 		}
+
```
Comment (marun) [inline pkg/kubefed2/enable/enable.go]: Please remove the inserted newlines - unrelated and nonfunctional formatting changes do not belong in this PR.

File: test/common/crudtester.go [hunk 17]
```diff
@@ -195,7 +195,10 @@ func (c *FederatedTypeCrudTester) CheckUpdate(fedObject *unstructured.Unstructur
 			}
 			clusterOverrides[key] = value
 		}
-		util.SetOverrides(obj, overrides)
+
+		if err := util.SetOverrides(obj, overrides); err != nil {
+			c.tl.Fatalf("Unexpected error: %s", err.Error())
```
Comment (marun) [inline test/common/crudtester.go]: Please prefer `"%v", err`

File: pkg/kubefed2/disable.go (hunk 2 above)
Comment (draveness) [reply pkg/kubefed2/disable.go]: > I have already disabled the check in this golangci file.

The struct and its vars are all private and not used directly with the form - `disableTypeOptions.xxx`, so the golint complains about that, though it was used in `disableType`. I think `disableTypeOptions` isn't useful and seems quite redundant, move all its vars and method into `disableType` still works the same. But we could keep that if you prefer that way.

File: pkg/kubefed2/enable/enable.go (hunk 16 above)
Comment (draveness) [reply pkg/kubefed2/enable/enable.go]: removed

**Comment by draveness** [comment]
Hi @marun, thanks for the review. And I just resolved all the conversations.

File: .golangci.yml [hunk 18]
```diff
@@ -0,0 +1,9 @@
+issues:
+  exclude-rules:
+    - path: pkg/kubefed2/enable/enable.go
+      linters:
+        - unused
+        - structcheck
+    - path: pkg/kubefed2/disable.go
+      linters:
+        - structcheck
```
Comment (shashidharatd) [inline .golangci.yml]: (suggestion) I would wish there are no exceptions and all code follow similar pattern

Comment (draveness) [reply .golangci.yml]: I came up with commit to remove the exceptions in golangci.yml, maybe you could take a look. https://github.com/kubernetes-sigs/federation-v2/pull/790/commits/9634a69c28cafaef091df727297d9b3e3d2c68e3

File: pkg/kubefed2/disable.go (hunk 2 above)
Comment (marun) [reply pkg/kubefed2/disable.go]: `disableTypeOptions` is intended to isolate options that will be set via flags.  Where there are values that are supplied via flags (`targetName` in this case), I am supportive of them being added to the command type (i.e. `disableType`).  Would that satisfy the golint check?

File: pkg/controller/util/planner/planner.go [hunk 19]
```diff
@@ -71,25 +72,41 @@ func (p *Planner) Plan(availableClusters []string, currentReplicaCount map[strin
...
-		}
+		}, nil
 	}
 
 	for _, cluster := range availableClusters {
 		if localRSP, found := p.preferences.Spec.Clusters[cluster]; found {
-			preferences = append(preferences, named(cluster, localRSP))
+			preference, err := named(cluster, localRSP)
+			if err != nil {
+				glog.Errorf("Unexpected err: %v\n", err)
```
Comment (marun) [inline pkg/controller/util/planner/planner.go]: Looking at this in a wider context this function should return the error rather than logging and continuing.  My apologies for not figuring this out sooner.

cc: @irfanurrehman

File: pkg/controller/util/planner/planner.go [hunk 20]
```diff
@@ -71,25 +72,41 @@ func (p *Planner) Plan(availableClusters []string, currentReplicaCount map[strin
...
+				continue
+			}
+
+			preferences = append(preferences, preference)
 		} else {
 			if localRSP, found := p.preferences.Spec.Clusters["*"]; found {
-				preferences = append(preferences, named(cluster, localRSP))
+				preference, err := named(cluster, localRSP)
+				if err != nil {
+					glog.Errorf("Unexpected err: %v\n", err)
```
Comment (marun) [inline pkg/controller/util/planner/planner.go]: ditto

File: pkg/kubefed2/disable.go [hunk 21]
```diff
@@ -62,17 +62,13 @@ var (
 
 type disableType struct {
 	options.GlobalSubcommandOptions
-	disableTypeOptions
```
Comment (marun) [inline pkg/kubefed2/disable.go]: As per my previous comment, I would prefer that the options set by flag remain in the type for consistency with the other kubefed commands.  We can revisit this for all commands in the future.

Comment (draveness) [reply pkg/kubefed2/disable.go]: OK, I'll revert the latest commit

File: pkg/controller/util/planner/planner.go (hunk 19 above)
Comment (draveness) [reply pkg/controller/util/planner/planner.go]: Please take a look at the latest commit, it returns errors in both method - `plan` and `schedule`

**Comment by marun** [comment]
@draveness lgtm.  Please squash to one commit and I'll approve.

**Comment by draveness** [comment]
> @draveness lgtm. Please squash to one commit and I'll approve.

done

**Comment by marun** [comment]
Nice work, thank you!

/approve
/lgtm

**Comment by draveness** [comment]
Hi @xunpan @marun @shashidharatd, the previous lgtm label was removed because of code rebasing, PTAL. :)
